src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,525 @@ void AsyncRPCOperation_sendmany::main() {
+                shieldedRecipients += 1;
+                if (isfromsprout_ && !allowRevealedAmounts_) {
+                    throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
+                            "Sending between shielded pools is not enabled by default because it will "
+                            "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                            "Resubmit with the `allowRevealedAmounts` parameter set to `true` if "
+                            "you wish to allow this transaction to proceed anyway.");
+                }
+            },
+            [&](const libzcash::UnifiedAddress& ua) {
+                // unreachable; currently disallowed by checks at construction
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Sending to unified addresses is disabled.");
+            }
+        }, recipient.address);
     }
 
     CAmount sendAmount = txValues.z_outputs_total + txValues.t_outputs_total;
-    txValues.targetAmount = sendAmount + minersFee;
-
-    // When spending coinbase utxos, you can only specify a single zaddr as the change must go somewhere
-    // and if there are multiple zaddrs, we don't know where to send it.
-    if (isfromtaddr_) {
-        // Only select coinbase if we are spending from a single t-address to a single z-address.
-        if (!useanyutxo_ && isSingleZaddrOutput) {
-            bool b = find_utxos(true, txValues);
-            if (!b) {
-                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient transparent funds, no UTXOs found for taddr from address.");
-            }
+    txValues.targetAmount = sendAmount + fee_;
+
+    builder_.SetFee(fee_);
+
+    // Only select coinbase if we are spending from at most a single t-address.
+    bool allowTransparentCoinbase =
+        !IsFromAnyTaddr(paymentSource_) && // allow coinbase inputs from at most a single t-addr
+        transparentRecipients == 0; // cannot send transparent coinbase to transparent recipients
+
+    // Set the dust threshold so that we can select enough inputs to avoid
+    // creating dust change amounts.
+    CAmount dustThreshold{DefaultDustThreshold()};
+
+    // Find spendable inputs, and select a minimal set of them that
+    // can supply the required target amount.
+    auto spendable = FindSpendableInputs(allowTransparentCoinbase);
+    if (!spendable.LimitToAmount(txValues.targetAmount, dustThreshold)) {
+        CAmount changeAmount{spendable.Total() - txValues.targetAmount};
+        if (changeAmount > 0 && changeAmount < dustThreshold) {
+            // TODO: we should provide the option for the caller to explicitly
+            // forego change (definitionally an amount below the dust amount)
+            // and send the extra to the recipient or the miner fee to avoid
+            // creating dust change, rather than prohibit them from sending
+            // entirely in this circumstance.
```

Comment:
```suggestion
            // entirely in this circumstance.
            // (Daira disagrees since this may leak note values.)
```

Reply:
I will add this comment to the branch that integrates UAs, so as to avoid a CI run.